frontend/controller/serviceworkers/sw-primary.js

```diff
@@ -1,12 +1,156 @@
+  'chelonia/private/invoke',
+  'state/vuex/state',
+  'state/vuex/getters',
+  'state/vuex/settings',
+  'gi.db/settings/save',
+  'gi.db/logs/save'
+].reduce(reducer, {})
+sbp('sbp/filters/global/add', (domain, selector, data) => {
+  if (domainBlacklist[domain] || selectorBlacklist[selector]) return
+  console.debug(`[sbp] ${selector}`, data)
+})
+
+// These are all of the events that will be forwarded to all open tabs and windows
+;[
+  CHELONIA_RESET, CONTRACTS_MODIFIED, CONTRACT_IS_SYNCING, EVENT_HANDLED, LOGIN,
+  LOGIN_ERROR, LOGOUT, ACCEPTED_GROUP, CHATROOM_USER_STOP_TYPING,
+  CHATROOM_USER_TYPING, DELETED_CHATROOM, LEFT_CHATROOM, LEFT_GROUP,
+  JOINED_CHATROOM, JOINED_GROUP, KV_EVENT, MESSAGE_RECEIVE, MESSAGE_SEND,
+  NAMESPACE_REGISTRATION, NEW_CHATROOM_UNREAD_POSITION, NEW_LAST_LOGGED_IN,
+  NEW_PREFERENCES, NEW_UNREAD_MESSAGES, NOTIFICATION_EMITTED,
+  NOTIFICATION_REMOVED, NOTIFICATION_STATUS_LOADED, OFFLINE, ONLINE,
+  PROPOSAL_ARCHIVED, SERIOUS_ERROR, SWITCH_GROUP
+].forEach(et => {
+  sbp('okTurtles.events/on', et, (...args) => {
+    const { data } = serializer(args)
+    const message = {
+      type: 'event',
+      subtype: et,
+      data
+    }
+    self.clients.matchAll()
+      .then((clientList) => {
+        clientList.forEach((client) => {
+          client.postMessage(message)
+        })
+      })
+  })
+})
+
+// Logs are treated especially to avoid spamming logs with event emitted
+// entries
+sbp('okTurtles.events/on', CAPTURED_LOGS, (...args) => {
+  const { data } = serializer(args)
+  const message = {
+    type: CAPTURED_LOGS,
+    data
+  }
+  self.clients.matchAll()
+    .then((clientList) => {
+      clientList.forEach((client) => {
+        client.postMessage(message)
+      })
+    })
+})
+
+sbp('sbp/selectors/register', {
+  'state/vuex/state': () => sbp('chelonia/rootState'),
+  'state/vuex/getters': (() => {
+    // Singleton lazily generated getters
+    let obj
+    return () => {
+      if (!obj) {
+        obj = Object.create(null)
+        Object.defineProperties(obj, Object.fromEntries(Object.entries(getters).map(([getter, fn]: [string, Function]) => {
+          return [getter, {
+            get: () => {
+              const state = sbp('chelonia/rootState')
+              return fn(state, obj)
+            }
+          }]
+        })))
+        Object.defineProperties(obj, Object.fromEntries(Object.entries(chatroomGetters).map(([getter, fn]: [string, Function]) => {
+          return [getter, {
+            get: () => {
+              const state = sbp('chelonia/rootState')
+              // `state.chatroom` represents the `chatroom` module. For the SW,
+              // this is defined in `sw-primary.js`.
+              return fn(state.chatroom || {}, obj, state)
+            }
+          }]
+        })))
+      }
+
+      return obj
+    }
+  })()
+})
+
+sbp('sbp/selectors/register', {
+  'controller/router': () => {
+    return { options: { base: '/app/' } }
+  }
```

Comment:
- Is this everything that should be returned?
- Can a comment be added here explaining why this is here?
- Can `/app/` be gotten programmatically (in case it ever changes)?

Reply:
> Is this everything that should be returned?

Yes.

> Can a comment be added here explaining why this is here?

Sure

> Can /app/ be gotten programmatically (in case it ever changes)?

No, it's only available to Vue, and it's hardcoded there too. The only way to pass it programmatically would be to pass it as a parameter when the SW is created, like `?path=/app`.

Reply:
> The only way to pass it programmatically would be to pass it as a parameter when the SW is created, like ?path=/app

Maybe we should do that...